### PR TITLE
2.0/feature log in sign up/backend

### PR DESCRIPTION
**Pull request by Oleh-Kril**

### CODE REVIEW

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 1]
```diff
@@ -63,7 +66,7 @@ public async Task<IActionResult> RefreshAsync()
         {
             var refreshToken = Request.Cookies["refreshToken"];
             
-            if (refreshToken == null || !_jwtService.ValidateToken(refreshToken))
+            if (refreshToken == null || string.IsNullOrEmpty(_jwtService.ValidateToken(refreshToken)))
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:69]: ValidateToke should return bool

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 2]
```diff
@@ -12,12 +13,15 @@ public class AuthController : ControllerBase
 {
     private readonly IUserService _userService;
     private readonly IJwtService _jwtService;
+    private readonly IEmailService _emailService;
     private readonly ILogger<AuthController> _logger;
-    public AuthController(ILogger<AuthController> logger, IUserService userService, IJwtService jwtService)
+    public AuthController(ILogger<AuthController> logger, IUserService userService,
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:18]: formating

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 3]
```diff
@@ -106,7 +109,7 @@ public async Task<IActionResult> LogOutAsync()
 
         try
         {
-            if(!_jwtService.ValidateToken(refreshToken))
+            if(string.IsNullOrEmpty(_jwtService.ValidateToken(refreshToken)))
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:112]: redundant

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 4]
```diff
@@ -120,5 +123,117 @@ public async Task<IActionResult> LogOutAsync()
             return BadRequest(ex.Message);
         }
     }
+    
+    [HttpGet("activate/{token}")]
+    [AllowAnonymous]
+    public async Task<IActionResult> ActivateAccountAsync(string token)
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:129]: to activate account i would not use token probably

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 5]
```diff
@@ -120,5 +123,117 @@ public async Task<IActionResult> LogOutAsync()
...
+    {
+        try
+        {
+            string userId = _jwtService.ValidateToken(token);
+            
+            if(!string.IsNullOrEmpty(userId))
+            {
+                await _userService.ActivateUserAccountAsync(userId);
+                return Redirect("http://localhost:3000/log-in");
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:138]: move addres to config file

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 6]
```diff
@@ -120,5 +123,117 @@ public async Task<IActionResult> LogOutAsync()
...
+    [AllowAnonymous]
+    public async Task<IActionResult> ActivateAccountAsync(string token)
+    {
+        try
+        {
+            string userId = _jwtService.ValidateToken(token);
+            
+            if(!string.IsNullOrEmpty(userId))
+            {
+                await _userService.ActivateUserAccountAsync(userId);
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:137]: Maybe redirtect if you sussfully activated account

File: SportHub.Business/Services/EmailService.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        
+        public async Task SendActivationEmailAsync(string userEmail, string activationLink)
+        {
+            var message = new MailMessage(_senderEmail, userEmail)
+            {
+                Subject = "Activate your account",
+                IsBodyHtml = true,
+                Body = 
+                    @"<body classname=""body-dark"" style=""margin: 0; padding: 0; box-sizing: border-box; background: #F9F9FB; font-family: 'Open Sans', sans-serif;"">
```
Comment (XashDM) [inline SportHub.Business/Services/EmailService.cs:32]: Make this a embeded resource

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 8]
```diff
@@ -120,5 +136,89 @@ public async Task<IActionResult> LogOutAsync()
...
+    [AllowAnonymous]
+    public async Task<IActionResult> ActivateAccountAsync(string token)
+    {
+        try
+        {
+            var isActivated = await _userService.ActivateUserAccountAsync(token);
+            
+            if (isActivated)
+            {
+                return Redirect("http://localhost:3000/log-in");
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:150]: localhost:3000 в конфіг

File: SportHub.API/Controllers/AuthorizationController.cs [hunk 9]
```diff
@@ -120,5 +136,89 @@ public async Task<IActionResult> LogOutAsync()
...
+        }
+    }
+    
+    [HttpGet("changePassword")]
+    [AllowAnonymous]
+    public async Task<IActionResult> ChangePasswordAsync(string token, string password)
+    {
+        try
+        {
+            ResponseWithBoolAndMessage response = await _userService.ChangePasswordAsync(token, password);
```
Comment (XashDM) [inline SportHub.API/Controllers/AuthorizationController.cs:168]: ResponseWithBoolAndMessage  have you considered my suggestion to use tuple

File: SportHub.API/Controllers/UserController.cs [hunk 10]
```diff
@@ -62,23 +61,27 @@ public async Task<IActionResult> GetByIdAsync([FromRoute] string id)
...
+            var existingUser = await _usersService.GetUserByIdAsync(newUserDto.UserId);
+            
+            if (existingUser == null)
+            {
+                return NotFound($"User with ID '{newUserDto.UserId}' not found.");
+            }
+            
+            User newUser = _mapper.Map<UserUpdateRequestDto, User>(newUserDto, existingUser);
+            
+            await _usersService.UpdateUserAsync(newUser);
```
Comment (XashDM) [inline SportHub.API/Controllers/UserController.cs:83]: check if user exists i can put inside UpdateUserAsync

File: SportHub.API/Controllers/GoogleAuthorizationController.cs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                }
+
+                var jwtResponse = await _jwtService.GenerateTokensAsync(userFromDb);
+                UserResponseDto userDto =  _mapper.Map<User, UserResponseDto>(jwtResponse.User);
+                
+                return Ok(new
+                {
+                    AccessToken = jwtResponse.AccessToken,
+                    User = userDto
+                });
```
Comment (XashDM) [inline SportHub.API/Controllers/GoogleAuthorizationController.cs:57]: Add to do so we can refactor and simplify this

File: SportHub.Business/Services/JwtService.cs [hunk 12]
```diff
@@ -29,25 +29,37 @@ public async Task<JwtResponse> GenerateTokensAsync(User user)
...
+                var tokenData = tokenHandler.ValidateToken(token,
+                    CreateTokenValidationParameters(),
+                    out SecurityToken validatedToken);
+                string response = tokenData.Claims.First(claim => claim.Type == ClaimTypes.NameIdentifier).Value;
+
+                return response;
+            }
+            catch
+            {
+                return "";
```
Comment (XashDM) [inline SportHub.Business/Services/JwtService.cs:59]: It's good idea to add logging here

File: SportHub.IntegrationTests/AuthorizationControllerTests.cs [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+            _testUserRequestDto = new UserRequestDto
+            {
+                FirstName = "testUserFirstName", 
+                LastName = "testUserLastName", 
+                Email = "testUserEmail@gmail.com", 
+                Password = "testUserPassword",
+            };
+        }
+
+        [Test, Order(1)]
```
Comment (XashDM) [inline SportHub.IntegrationTests/AuthorizationControllerTests.cs:74]: It's bad practice to have order - remind me to discuss it on next meeting

**Review by XashDM — APPROVED** [review]
Переглянь коментарі

File: SportHub.IntegrationTests/AuthorizationControllerTests.cs (hunk 13 above)
Comment (Oleh-Kril) [reply SportHub.IntegrationTests/AuthorizationControllerTests.cs:74]: I need order to be sure that inserting is before activation user account and so on

**Pull request merged**